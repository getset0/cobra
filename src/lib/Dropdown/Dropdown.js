import React from "react";
import styled, { css } from "styled-components";
import PropTypes from "prop-types";
import { theme } from "../constants";

const DropdownWrapper = styled.div`
  width: ${props => `${props.width}px`};
  max-width: 180px;
  border-radius: 3px;
  transition: all .3s ease-out;
  z-index: 99;
  ${props => props.withBorder && (
    props.isOpen ? css`
      border: 1px solid #D8D8D8;
      box-shadow: 1px 2px 3px #D8D8D8;
      background: ${props.theme.whiteColor};
    ` : css`
      border: 1px solid transparent;
      background: transparent;
      box-shadow: 1px 2px 3px transparent;
  `)}
`;

class Dropdown extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      isOpen: this.props.isOpen || false
    };
  }
  handleTriggerClick = () => {
    this.setState((prevState) => ({ isOpen: !prevState.isOpen }),  () =>
      this.props.onStateChange && this.props.onStateChange(this.state.isOpen)
    );
  }
  render() {
    return (
      <DropdownWrapper
        isOpen={this.state.isOpen}
        withBorder={this.props.withBorder}
        width={this.props.width}
      >
        {
          this.props.children({
            isOpen: this.state.isOpen,
            handleTriggerClick: this.handleTriggerClick
          })
        }
      </DropdownWrapper>
    );
  }
}

Dropdown.propTypes = {
  withBorder: PropTypes.bool,
  isOpen: PropTypes.bool,
  width: PropTypes.number,
  theme: PropTypes.object,
  onStateChange: PropTypes.func,
  children: PropTypes.any.isRequired
};

Dropdown.defaultProps = {
  withBorder: true,
  isOpen: false,
  width: 130,
  theme: theme
};

export default Dropdown;


