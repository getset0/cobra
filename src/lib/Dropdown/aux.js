import styled, { css } from "styled-components";
import PropTypes from "prop-types";
import { theme } from "../constants";

export const DropdownTrigger = styled.div`
  cursor: pointer;
  padding: 10px;
`;

export const DropdownButton = styled.div`
  display: flex;
  justify-content: flex-end;
  ${props => props.isOpen && css`
    color: ${props => props.theme.primaryColor}
  `}
`;

DropdownButton.propTypes = {
  isOpen: PropTypes.bool,
  theme: PropTypes.object
};

DropdownButton.defaultProps = {
  theme: theme
};

export const DropdownBody = styled.div`
  transition: all .3s ease-out;
  overflow: hidden;
  width: 100%;
  height: 100%;
  z-index: 99;
  background: ${props => props.theme.whiteColor};
  ${props => props.isOpen ? css`
    max-height: 200px;
  ` : css`
    max-height: 0;
  `}
`;

DropdownBody.propTypes = {
  isOpen: PropTypes.bool,
  theme: PropTypes.object,
  children: PropTypes.any.isRequired
};

DropdownBody.defaultProps = {
  theme: theme
};
