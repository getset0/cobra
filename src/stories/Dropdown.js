import React from "react";
import { storiesOf } from "@storybook/react";
import { withInfo } from "@storybook/addon-info";
import { withBackgrounds } from "@storybook/addon-backgrounds";
import { backgroundOptions } from "./util";
import {
  Dropdown,
  DropdownBody,
  DropdownTrigger,
  DropdownButton,
  List,
  ListItem
} from "../lib";

storiesOf("Dropdown", module)
  .addDecorator(withInfo)
  .addDecorator(withBackgrounds(backgroundOptions))
  .add("Dropdown", () => (
    <Dropdown withBorder isOpen>
      {({ isOpen, handleTriggerClick }) => (
        <React.Fragment>
          <DropdownTrigger onClick={handleTriggerClick}>
            <DropdownButton isOpen={isOpen}>Trigger</DropdownButton>
          </DropdownTrigger>
          <DropdownBody isOpen={isOpen}>
            <List>
              <ListItem onClick={() => console.log("item1")}>item1</ListItem>
              <ListItem onClick={() => console.log("item2")}>item2</ListItem>
              <ListItem onClick={() => console.log("item3")}>item3</ListItem>
              <ListItem onClick={() => console.log("item4")}>item4</ListItem>
            </List>
          </DropdownBody>
        </React.Fragment>
      )}
    </Dropdown>
  ));

