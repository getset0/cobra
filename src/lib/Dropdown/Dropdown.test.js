import React from "react";
import Dropdown from "./Dropdown";
import {
  DropdownTrigger,
  DropdownBody
} from "./aux";
import "jest-styled-components";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";

const BasicDropdown = (props) =>
  <Dropdown withBorder {...props}>
    {({ isOpen, handleTriggerClick }) => (
      <React.Fragment>
        <DropdownTrigger onClick={handleTriggerClick}>
          Trigger
        </DropdownTrigger>
        <DropdownBody isOpen={isOpen}>
          Body
        </DropdownBody>
      </React.Fragment>
    )}
  </Dropdown>;

it("renders Dropdown without crashing", () => {
  shallow(BasicDropdown());
});

it("renders Dropdown correctly", () => {
  const wrapper = shallow(BasicDropdown());
  const tree = toJson(wrapper);
  expect(tree).toMatchSnapshot();
});

it("test Dropdown toggle", () => {
  const mockStateChange = jest.fn(x => x);
  const wrapper = shallow(
    BasicDropdown({ onStateChange: mockStateChange })
  );
  // has a trigger
  const ddTrigger = wrapper.find(DropdownTrigger);
  expect(ddTrigger).toHaveLength(1);

  // click 3 times on trigger
  ddTrigger.prop("onClick")();
  ddTrigger.prop("onClick")();
  ddTrigger.prop("onClick")();

  // expect to have called mock 3 times
  expect(mockStateChange.mock.calls).toHaveLength(3);

  // expect mock params to be consistent
  expect(mockStateChange.mock.calls[0][0]).toBeTruthy();
  expect(mockStateChange.mock.calls[1][0]).toBeFalsy();
  expect(mockStateChange.mock.calls[2][0]).toBeTruthy();
});
