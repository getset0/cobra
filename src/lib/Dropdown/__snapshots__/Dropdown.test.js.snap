// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Dropdown correctly 1`] = `
<styled.div
  isOpen={false}
  width={130}
  withBorder={true}
>
  <styled.div
    onClick={[Function]}
  >
    Trigger
  </styled.div>
  <styled.div
    isOpen={false}
    theme={
      Object {
        "blackColor": "#000000",
        "darkBlueColor": "#02469B",
        "darkGrayColor": "#4A4A4A",
        "darkestBlueColor": "#02254D",
        "darkestGrayColor": "#303030",
        "greenColor": "#76B041",
        "lightBlueColor": "#0190D0",
        "lightGrayColor": "#D8D8D8",
        "lightestBlueColor": "#00FEFE",
        "mediumGrayColor": "#979797",
        "primaryColor": "#1676B7",
        "secondaryColor": "#17BEBB",
        "whiteBackgroundColor": "#fdfdfd",
        "whiteColor": "#ffffff",
        "yellowColor": "#FFC914",
      }
    }
  >
    Body
  </styled.div>
</styled.div>
`;
